### PR TITLE
Add OPA policy support

### PR DESCRIPTION
**Pull request by binodmx**
## Purpose
This PR resolves https://github.com/wso2/apk/issues/1164 by adding OPA policy config for API Policies. An OPA policy can be added by defining `oPAPolicy` property in APIPolicy kind. Following is an sample APIPolicy kind which includes a OPA policy.

```yaml
apiVersion: dp.wso2.com/v1alpha1
kind: APIPolicy
metadata:
  name: sample-api-policy
  namespace: apk
spec:
  override:
    oPAPolicy:
      serverURL: http://opa-svc.apk:8181/v1/data
      policy: opa/test
      rule: allow
  targetRef:
    group: dp.wso2.com
    kind: API
    name: http-bin-api
    namespace: apk
```

### CODE REVIEW

File: adapter/internal/operator/apis/dp/v1alpha1/apipolicy_types.go [hunk 1]
```diff
@@ -78,6 +79,33 @@ type CORSPolicy struct {
 	AccessControlExposeHeaders    []string `json:"accessControlExposeHeaders,omitempty"`
 }
 
+// OPAPolicy holds OPA policy information
+type OPAPolicy struct {
+	// ServerURL is the URL of the OPA server
+	ServerURL string `json:"serverURL,omitempty"`
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/operator/apis/dp/v1alpha1/apipolicy_types.go:85]: please add validation regex for url

File: gateway/enforcer/org.wso2.apk.enforcer/src/main/java/org/wso2/apk/enforcer/interceptor/opa/OPADefaultRequestGenerator.java [hunk 2]
```diff
@@ -96,15 +96,28 @@ public String generateRequest(String policyName, String rule, Map<String, String
...
+                    "Empty result received for the OPA policy " + policyName);
+        } else {
+            try {
+                JSONObject response = new JSONObject(opaResponse);
+                if (rule.isEmpty()) {
+                    JSONObject result = (JSONObject) response.get(OPAConstants.OPA_RESPONSE_RESULT_KEY);
+                    return result.getBoolean(OPAConstants.OPA_RESPONSE_DEFAULT_RULE);
+                } else {
+                    return response.getBoolean(OPAConstants.OPA_RESPONSE_RESULT_KEY);
+                }
```
Comment (AmaliMatharaarachchi) [inline gateway/enforcer/org.wso2.apk.enforcer/src/main/java/org/wso2/apk/enforcer/interceptor/opa/OPADefaultRequestGenerator.java:113]:
```suggestion
                } 
                 return response.getBoolean(OPAConstants.OPA_RESPONSE_RESULT_KEY);
                
```

File: adapter/internal/operator/apis/dp/v1alpha1/apipolicy_types.go (hunk 1 above)
Comment (binodmx) [reply adapter/internal/operator/apis/dp/v1alpha1/apipolicy_types.go:86]: Added by https://github.com/wso2/apk/pull/1193/commits/cd5afd7d2c0907424415d64317afb9ad9a5dc707

File: gateway/enforcer/org.wso2.apk.enforcer/src/main/java/org/wso2/apk/enforcer/interceptor/opa/OPADefaultRequestGenerator.java (hunk 2 above)
Comment (binodmx) [reply gateway/enforcer/org.wso2.apk.enforcer/src/main/java/org/wso2/apk/enforcer/interceptor/opa/OPADefaultRequestGenerator.java]: Resolved by https://github.com/wso2/apk/pull/1193/commits/cd5afd7d2c0907424415d64317afb9ad9a5dc707

**Comment by AmaliMatharaarachchi** [comment]
we will be reopening this PR with the mediation app implementation. Hence we will be closing as we are expecting changes to these CRs in future

**Pull request closed without merging**